src/data/real/basic.lean

```diff
@@ -48,64 +87,104 @@ instance : star_ring ℝ          := star_ring_of_comm
 
 /-- Coercion `ℚ` → `ℝ` as a `ring_hom`. Note that this
 is `cau_seq.completion.of_rat`, not `rat.cast`. -/
-def of_rat : ℚ →+* ℝ := ⟨of_rat, rfl, of_rat_mul, rfl, of_rat_add⟩
+def of_rat : ℚ →+* ℝ :=
+by refine_struct { to_fun := of_cauchy ∘ of_rat };
+  simp [of_rat_one, of_rat_zero, of_rat_mul, of_rat_add,
+    one_cauchy, zero_cauchy, ← mul_cauchy, ← add_cauchy]
+
+lemma of_rat_apply (x : ℚ) : of_rat x = of_cauchy (cau_seq.completion.of_rat x) := rfl
 
 /-- Make a real number from a Cauchy sequence of rationals (by taking the equivalence class). -/
-def mk (x : cau_seq ℚ abs) : ℝ := cau_seq.completion.mk x
+def mk (x : cau_seq ℚ abs) : ℝ := ⟨cau_seq.completion.mk x⟩
 
-theorem of_rat_sub (x y : ℚ) : of_rat (x - y) = of_rat x - of_rat y :=
-congr_arg mk (const_sub _ _)
+theorem mk_eq {f g : cau_seq ℚ abs} : mk f = mk g ↔ f ≈ g :=
+ext_cauchy_iff.trans mk_eq
 
-instance : has_lt ℝ :=
-⟨λ x y, quotient.lift_on₂ x y (<) $
+@[irreducible]
+private def lt : ℝ → ℝ → Prop | ⟨x⟩ ⟨y⟩ :=
+quotient.lift_on₂ x y (<) $
   λ f₁ g₁ f₂ g₂ hf hg, propext $
   ⟨λ h, lt_of_eq_of_lt (setoid.symm hf) (lt_of_lt_of_eq h hg),
-   λ h, lt_of_eq_of_lt hf (lt_of_lt_of_eq h (setoid.symm hg))⟩⟩
+   λ h, lt_of_eq_of_lt hf (lt_of_lt_of_eq h (setoid.symm hg))⟩
```

Comment:
How would this definition be ported to Lean 4? It looks like you're using that `real` is (definitionally) a wrapper around Cauchy sequences.

Reply:
Oh, to answer my own question: I think the idea is that you already have an instance `conditionally_complete_linear_ordered_field` on the reals, so that the order is already defined beforehand. Never mind then.

Reply:
The idea here is that we can do `@[irreducible] def lt` in Lean 4 as well.  (By simulating it using a constant using an `irreducible_def` macro as described on zulip.)  What we can't do is have `attribute [irreducible]` at the end of the file.

The `conditionally_complete_linear_ordered_field` approach would also work, but this is not the plan for this PR (or necessary for the port).

Reply:
Oh, if you're not doing the `conditionally_complete_linear_ordered_field` approach, then I don't understand it anymore. When you write
```
def lt : ℝ → ℝ → Prop | ⟨x⟩ ⟨y⟩ :=
quotient.lift_on₂ x y (<) ...
```
you are using that `ℝ` is a specifically defined thing. This fails if `ℝ` is a constant. How do you do this in Lean 4?

Reply:
So the reason why we put irreducible on these definitions in Lean 3 is because of performance issues.  We want Lean to never unfold them.

This *particular mechanism* of preventing reduction by adding the irreducible attribute at the end of a file is no longer available in Lean 4.  However other mechanisms can be used instead.  First, we can use structures.  Structures do not reduce (more precisely, the type of the structure doesn't).  The second (new) one is constants: if you define something using `constant foo := bar` then foo also doesn't reduce, but we have no idea that `foo` was defined as `bar`.  By writing `constant foo : { x // x = bar } := ⟨bar, rfl⟩` we package the definition into the type.  This way, we can simulate the Lean 3 construct `@[irreducible] def foo := bar` in Lean 4.  (Thanks to macros this can be used as a drop-in replacement.)

The question that arose on Zulip is whether such a bundling is problematic.  For example, Sebastien feared that if we can never use the definitional equality for a constant, then we would end up in DTT hell.

This PR answers this question negatively in the case of `real`. For other applications of irreducible, like `opposite` or `with_top` and co., it is still unclear if we can easily replace irreducible with the mechanisms available in Lean 4.

(I didn't want to cause any confusion by mentioning the axiomatic approach with `conditionally_complete_linear_ordered_field`.  That would also be interesting, but I'll let somebody else revive #3292.  The point of this PR is to study general-purpose replacements of the irreducible attribute.  Even if we can do it for the reals in a mathematically more pleasing way, that doesn't help us in other places.)

Reply:
Ah, got it. So in this PR you're just showing that it's fine to make `real` a structure, without breaking the whole library. You're not yet proposing a specific way to define `Real` in Lean 4.

Thanks for clearing up my confusion!